back/src/main/java/com/ubikloadpack/jmeter/ulp/observability/util/ErrorsMap.java

```diff
@@ -0,0 +1,119 @@
+package com.ubikloadpack.jmeter.ulp.observability.util;
+
+import java.util.Map;
+import java.util.OptionalDouble;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+
+/**
+ * The ErrorsMap stores for each error type it's occurrences.
+ * The occurrences of an error type is handled by {@link ErrorTypeInfo}
+ */ 
+public class ErrorsMap {
+	private ConcurrentHashMap<String, ErrorTypeInfo> errorsPerType;
+	
+	public ErrorsMap() {
+		this(new ConcurrentHashMap<>());
+	}
+
+
+	public ErrorsMap(ConcurrentHashMap<String, ErrorTypeInfo> errorsPerType) {
+		super();
+		this.errorsPerType = errorsPerType;
+	}
+
+
+
+	/**
+	 * Add an error type and increment it's occurrence. If the error type already exists
+	 * so it's occurrence will be incremented.
+	 * @param errorType the error type to add.
+	 * @return The information of the added error type.
+	 */
+	public ErrorTypeInfo addErrorTypeAndCount(String errorType) {
+		ErrorTypeInfo errorTypeInfo = this.errorsPerType.computeIfAbsent(errorType, s -> new ErrorTypeInfo(errorType));
+	    errorTypeInfo.increment();
+	    return errorTypeInfo;
+	}
+	
+	/**
+	 * Compute the frequency of that error type.
+	 * @param errorType the type of error for which the frequency will be calculated
+	 * @param errorsTotal the total errors. Should get it's value from the Micrometer registry.
+	 * @return an optional that contains the percentage of occurrences of a specific error type 
+	 * out of the total number of errors. Returns an empty optional if no errorType key was found.
+	 */
+	public OptionalDouble computeErrorTypeFrequency(String errorType, Long errorsTotal) {
+		if (this.errorsPerType.contains(errorType)) {
+			return OptionalDouble.of(errorsPerType.get(errorType).computeErrorRateAmongErrors(errorsTotal));
+		}
+		return OptionalDouble.empty();
+	}
+	
+	/**
+	 * Compute error rate of that error type.
+	 * @param errorType the type of error for which the error rate will be calculated.
+	 * @param requestsTotal the total requests. Should get it's value from the Micrometer registry.
+	 * @return an optional that contains the percentage of queries that raised an error of this type. 
+	 * Returns an empty optional if no errorType key was found.
+	 */
+	public OptionalDouble computeErrorRateForType(String errorType, Long requestsTotal) {

```

Comment:
Why not use Optional<Double>

Reply:
They are roughly similar. The difference is that the OptionalDOuble holds the double value as a primitive type.

Reply:
ok keep it like this